### PR TITLE
CI Fixes

### PR DESCRIPTION
**Pull request by Judahmeek**
This PR upgrades Puma to v6 to make it compatible with Rack v3.

This PR also adds new workflows so that all workflows required by our repository settings actually run.

- - -
This change is [Reviewable](https://reviewable.io/reviews/shakacode/react_on_rails/1581)

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -1,13 +1,17 @@
...
+      - 'master'
+  pull_request:
 
 jobs:
   build-dummy-app-webpack-test-bundles:
     strategy:
       matrix:
         ruby: [2.7, 3.2]
-        node: [16, 18]
+        node: [14, 16, 18]
```
Comment (ahangarha) [inline .github/workflows/main.yml]: Is there any particular reason for bringing back Node 14?
Before we decided to drop support for Node 14. (Though it seems I missed it in some other places)

Comment (Judahmeek) [reply .github/workflows/main.yml]: If we dropped support for it, then we need to tell Justin to remove any workflows for Node v14 from the list of required workflows.

Screenshot 2023-11-14 at 10 26 22 AM

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -79,8 +83,8 @@ jobs:
     needs: build-dummy-app-webpack-test-bundles
     strategy:
       matrix:
-        ruby: [2.7]
-        node: [14]
+        ruby: [2.7, 3.2]
+        node: [14, 16, 18]
```
Comment (justin808) [inline .github/workflows/main.yml]: newer node should be added

File: .github/workflows/package-js-tests.yml [hunk 3]
```diff
@@ -1,12 +1,17 @@
...
+    branches:
+      - 'master'
+  pull_request:
 
 jobs:
   build:
     strategy:
       matrix:
-        node: [16, 18]
+        node: [14, 16, 18]
```
Comment (justin808) [inline .github/workflows/package-js-tests.yml]: newer node

File: .github/workflows/rspec-package-specs.yml [hunk 4]
```diff
@@ -1,13 +1,17 @@
...
+      - 'master'
+  pull_request:
 
 jobs:
   build:
     strategy:
       matrix:
         ruby: [2.7, 3.2]
-        node: [16, 18]
+        node: [14, 16, 18]
```
Comment (justin808) [inline .github/workflows/rspec-package-specs.yml]: same

**Review by justin808 — CHANGES_REQUESTED** [review]

**Review by ahangarha — APPROVED** [review]
I agree with this decision of testing with the oldest (that we care about) and the newest Node version.

**Review by justin808 — APPROVED** [review]

**Pull request merged**